// _datdot-service-helpers/feed-storage.js
const intercept = require('_datdot-service-helpers/intercept-hypercore-storage')
const brotli = require('_datdot-service-helpers/brotli')
const parse_decompressed = require('_datdot-service-helpers/parse-decompressed')

const ENCODED_DATA_PREFIX = 0
const DECODED_PREFIX = 1
const ENCODED_DATA_SIGNATURE_PREFIX = 2
const NODES_PREFIX = 3
const UNIQUE_EL_PREFIX = 4
var stuff = { storeEnc: [], storeDec: [], getEnc: [], getDec: [] }

module.exports = class FeedStorage {
  /**
  brotli is an object with `async encode(rawData)` and `async decode(encodedData)`
  db is a levelup instance
  feed is a hypercore instance that will be used to store / seed the hosted hypercore
  **/
  constructor ({ db, feed, log }) {
    this.db = db
    this.log = log
    this.feed = feed
    this.unintercept = intercept(feed, {
      getData: async (index) => this.getData(index),
      putData: async (index, data, offset) => this.putData(index, data, offset)
    })
  }

  // Invoked by the encoder so that the host will store the encoded data
  async storeEncoded ({ index, encoded_data_signature, encoded_data, unique_el, nodes }) {
    this.log({ type: 'hoster', data: { text: `Hoster storage - starting to store`, index, encoded_data: encoded_data.toString('hex') }})
    // Get the decoded data at the index
    // In parallel, decode the encoded data
    const decompressed = await brotli.decompress(encoded_data)
    const decoded = parse_decompressed(decompressed, unique_el)
    this.log({ type: 'storage', data: { text: `Getting decoded before storing`, decoded }})
    stuff.storeEnc[index] = encoded_data
    stuff.storeDec[index] = decoded

    // If it's the same save the encoded data
    // and delete the decoded data if it exists
    await this._putEncodedData(index, encoded_data)
    await this._delDecoded(index)
    await this._putEncodedDataSignature(index, encoded_data_signature)
    await this._putNodes(index, nodes) // TODO: store deduplicated nodes 
    await this._putUniqueEl(index, Buffer.from(unique_el, 'binary'))
    this.log({ type: 'hoster', data: [`Hoster storage - all stored for chunk ${index}`]})
  }

  // Invoked by whoever to test that the hoster is actually hosting stuff
  async getProofOfStorage (index) {
    const [encoded_data, encoded_data_signature, nodes] = await Promise.all([
      this._getEncodedData(index),
      this._getEncodedDataSignature(index),
      this._getNodes(index),
    ])
    return {
      feedKey: this.feed.key,
      index,
      encoded_data,
      encoded_data_signature,
      nodes,
    }
  }

  // Meant to be used by the hypercore storage
  async getData (index) {
    this.log({ type: 'storage', data: { text: 'getData intercepted', index }})
    var decoded
    try {
      const encoded_data = await this._getEncodedData(index)
      this.log({ type: 'storage', data : { text: 'got encoded data for', index, encoded_data } })
      // Got encoded data!
      // Decode and return it
      const enc = stuff.storeEnc[index]
      const dec = stuff.storeDec[index]
      // TODO: (maybe it's ok)
      const decompressed = await brotli.decompress(encoded_data)
      const unique_el = await this._getUniqueEl(index)
      decoded = parse_decompressed(decompressed, unique_el)
      return decoded
    } catch (error) {
      // Key doesn't exist? Try to get decoded version
      this.log({ type: 'Error', data: { text: 'Encoded is not yet stored', index }})
      try {
        // If it exists, return it
        decoded = await this._getDecoded(index)
        this.log({ type: 'storage', data: { text: 'got decoded version for', index, decoded: decoded.toString('hex') } })
        return decoded
        // If we got the value, send it to the CB
      } catch(error) {
        this.log({ type: 'Error', data: { text: 'Error getting decoded', index }})
        console.log({ text: 'error getting decoded', error})
        throw error
      }
    }
  }

  // Meant to be used by the hypercore storage
  async putData (index, data, offset) {
    this.log({ type: 'storage', data: { text: 'putData intercepted', index }})
    // Check if encoded data exists
    try {
      const decoded = await this._getDecoded(index)
      // We already have the encoded version so whatever
      this.log({ type: 'storage', data: { text: 'We already have the encoded version', index }})
      this.log({ type: 'storage', data: { text: 'Already have encoded, here is decoded', decoded }})
      return
    } catch (error) {
      // We don't have the encoded version so we should store the decoded version
      try {
        await this._putDecoded(index, data)
        this.log({ type: 'storage', data: { text: `We don't have the encoded version so we should store the decoded version`, index }})
        return
      } catch (err) {
        this.log({ type: 'storage', data: { text: `Error in putData`, index, err }})
        throw err
      }
    }
  }

  async _getEncodedData (index) {
    try {
      return this.db.get(makeKey(ENCODED_DATA_PREFIX, index))
    } catch(err) {
      this.log({ type: 'Error in _getEncodedData', data: err})
    }
  }
  async _getDecoded (index) {
    try {
      return this.db.get(makeKey(DECODED_PREFIX, index))
    } catch(err) {
      this.log({ type: 'Error in _getDecoded', data: err})
    }
  }
  async _getEncodedDataSignature (index) {
    try {
      return this.db.get(makeKey(ENCODED_DATA_SIGNATURE_PREFIX, index))
    } catch(err) {
      this.log({ type: 'Error in _getEncodedDataSignature', data: err})
    }
  }
  async _getNodes (index) {
    try {
      const nodesBuf = await this.db.get(makeKey(NODES_PREFIX, index))
      const json =  nodesBuf.toString('utf-8')
      const nodes = JSON.parse(json)
      return nodes.map(node => {
        node.hash = Buffer.from(node.hash, 'hex')
        return node
      })
    } catch(err) {
      this.log({ type: 'Error in _getNodes', data: err})
    }
  }
  async _getUniqueEl (index) {
    try {
      return this.db.get(makeKey(UNIQUE_EL_PREFIX, index))
    } catch(err) {
      this.log({ type: 'Error in _getUniqueEl', data: err})
    }
  }

  async _putEncodedData (index, data) {
    return this.db.put(makeKey(ENCODED_DATA_PREFIX, index), data, Buffer.from([0, 255]), err => {
      if (err) this.log({ type: 'fail', data: { text: '_putEncodedData', err } })
    })
  }
  async _putDecoded (index, data) {
    return this.db.put(makeKey(DECODED_PREFIX, index), data, Buffer.from([0, 255]), err => {
      if (err) this.log({ type:'fail', data: { text: '_putDecoded', err } })
    })
  }
  async _putEncodedDataSignature (index, proof) {
    return this.db.put(makeKey(ENCODED_DATA_SIGNATURE_PREFIX, index), proof, Buffer.from([0, 255]), err => {
      if (err) this.log({ type: 'fail', data: { text: '_putEncodedDataSignature', err } })
    })
  }
  async _putNodes (index, nodes) {
    nodes = Buffer.from(JSON.stringify(nodes), 'utf-8')
    return this.db.put(makeKey(NODES_PREFIX, index), nodes, Buffer.from([0, 255]), err => {
      if (err) this.log({ type: 'fail', data: { text: '_putNodes', err} })
    })
  }
  async _putUniqueEl (index, data) {
    return this.db.put(makeKey(UNIQUE_EL_PREFIX, index), data, Buffer.from([0, 255]), err => {
      if (err) this.log({ type: 'fail', data: { text: '_putUniqueEl', err} })
    })
  }

  async _delDecoded (index) {
    try {
      await this.db.del(makeKey(DECODED_PREFIX, index))
    } catch {
      // Whatever, it's probably not saved
    }
  }

  async close () {
    await this.feed.close()
    await this.discovery.destroy()
    await this.db.close()
  }
  
  async destroy (log) {
    log({ type: 'feed', data: { text: 'Destroying discovery and closing feed and db' }})
    await this.feed.close()
    await this.discovery.destroy()
    await this.db.clear()
    await this.db.close()
  }
}

function makeKey (type, index) {
  // Allocate a buffer that holds enough for the type and the index
  const typedKey = new Uint32Array(2)

  typedKey[0] = type
  typedKey[1] = index

  return Buffer.from(typedKey.buffer)
}
