// _datdot-service-helpers/try-refresh-discovery.js
module.exports = try_refresh_discovery

async function try_refresh_discovery ({ swarm, topic, tasks, log }) {
	return new Promise(async (resolve, reject) => {
		tasks = Object.assign({ 
			author: false, // { server: true }
			sponsor: false, // { client: true } 
			encoder2author: false, // { client: true }
			encoder2attester: false, // { server: true }
			attester2encoder: false, // { client: true }
			attester2hoster: false, // { server: true }
			hoster2author: false, // { server: true, client: true }
			hoster2attester: false, // { client: true }
			hoster: false, // { server: true }
			perf_attester: false,
			storage_attester: false,
			perf_hoster: false,
			storage_hoster: false
		}, tasks)

		const { 
			author, sponsor, encoder2author, encoder2attester, attester2encoder, attester2hoster, hoster2author, hoster2attester, hoster, perf_attester
		} = tasks 
		
		const any_client_only = sponsor || encoder2author || attester2encoder || hoster2attester 
		const no_client = !sponsor && !encoder2author && !attester2encoder && !hoster2attester
		const any_server_only = author || encoder2attester || attester2hoster || hoster
		const no_server = !author && !encoder2attester && !attester2hoster && !hoster
		const server_and_client = hoster2author
		
		var mode
		if (server_and_client) mode = { server: true, client: true }
		else if (any_client_only && any_server_only) mode = { server: true, client: true }
		else if (any_client_only && no_server) mode = { server: false, client: true }
		else if (no_client && any_server_only) mode = { server: true, client: false }
		
		try {
			const discovery = swarm.status(topic)
			log({ type: 'swarm', data: { modeee: mode, s: discovery.isServer, c: discovery.isClient, tasks, any_client_only,no_client, any_server_only, no_server, server_and_client } })
			if (discovery.isServer === mode.server && discovery.isClient === mode.client) return resolve(mode)
			discovery.refresh(mode)
			log({ type: 'swarm', data: { text: 'refresh discovery', topic: topic.toString('hex'), mode }})
			// console.log({ discovery: storage.discovery })
			resolve(mode)
		} catch (err) {
			log({ type: 'Error', data: { text: 'Error: in updating swarm discovery mode', err }})
			reject()
		}
	})
}
