// datdot-ui-dropdown/src/node_modules/make-button.js
const {i_button} = require('datdot-ui-button')
module.exports = make_button
function make_button ({page, option = {}, mode, expanded, theme = {}}, protocol) {
    const {flow = 'ui-dropdown', name, role = 'listbox', body, icons, cover, disabled = false} = option
    const match = mode.match(/single|multiple/)
    const button_mode = match ? 'selector' : 'menu'
    const {style = ``, props = {}, grid = {}} = theme

    return i_button({
        page,
        flow, 
        name, 
        role,
        body,
        icons,
        cover, 
        mode: button_mode, 
        expanded, disabled, 
        theme: {
            style: `
                :host(i-button) > .icon {
                    transform: rotate(0deg);
                    transition: transform 0.4s ease-in-out;
                }
                :host(i-button[aria-expanded="true"]) > .icon {
                    transform: rotate(${mode === 'single-select' ? '-180' : '0' }deg);
                }
                ${style}
            `,
            props,
            grid
        }
    }, protocol(name))
}
