// @polkadot/api-derive/elections/info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.info = info;

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _types = require("@polkadot/types");

var _util = require("../util");

// Copyright 2017-2019 @polkadot/api-derive authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
function deriveElections(api, candidates, members, runnersUp, candidacyBond, desiredSeats, termDuration, votingBond) {
  return {
    candidates,
    candidateCount: (0, _types.createType)(api.registry, 'u32', candidates.length),
    candidacyBond,
    desiredSeats,
    members: members.sort((a, b) => b[1].cmp(a[1])),
    runnersUp: runnersUp.sort((a, b) => b[1].cmp(a[1])),
    termDuration,
    votingBond
  };
}

function queryElections(api) {
  const section = api.query.electionsPhragmen ? 'electionsPhragmen' : 'elections'; // NOTE We have an issue where candidates can return `null` for an empty array, hence
  // we are not using multi queries here, so empty array is empty (instead of space-filled)

  return (0, _rxjs.combineLatest)([api.query[section].candidates(), api.query[section].members(), api.query[section].runnersUp()]).pipe((0, _operators.map)((_ref) => {
    let [candidates, members, runnersUp] = _ref;
    return deriveElections(api, candidates, members, runnersUp, api.consts[section].candidacyBond, api.consts[section].desiredMembers, api.consts[section].termDuration, api.consts[section].votingBond);
  }));
}
/**
 * @name info
 * @returns An object containing the combined results of the storage queries for
 * all relevant election module properties.
 * @example
 * <BR>
 *
 * ```javascript
 * api.derive.elections.info(({ members, candidates }) => {
 *   console.log(`There are currently ${members.length} council members and ${candidates.length} prospective council candidates.`);
 * });
 * ```
 */


function info(api) {
  return (0, _util.memo)(() => queryElections(api));
}
