// @polkadot/jsonrpc/account.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _method = _interopRequireDefault(require("./create/method"));

var _param = _interopRequireDefault(require("./create/param"));

// Copyright 2017-2019 @polkadot/jsonrpc authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// NOTE order here is the same as in Rust, alphabetical below
const nextIndex = {
  description: 'Retrieves the next accountIndex as available on the node',
  isOptional: true,
  params: [(0, _param.default)('accountId', 'AccountId')],
  type: 'Index'
};
const section = 'account';
/**
 * @summary Calls to account-specific information.
 */

var _default = {
  isDeprecated: false,
  isHidden: false,
  description: '(Optional) Methods that retrieves account-specific information',
  section,
  methods: {
    nextIndex: (0, _method.default)(section, 'nextIndex', nextIndex)
  }
};
exports.default = _default;
