// @polkadot/jsonrpc/contracts.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _method = _interopRequireDefault(require("./create/method"));

var _param = _interopRequireDefault(require("./create/param"));

// Copyright 2017-2019 @polkadot/jsonrpc authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// NOTE order here is the same as in Rust, alphabetical below
const call = {
  description: 'Executes a call to a contract',
  isOptional: true,
  params: [(0, _param.default)('callRequest', 'ContractCallRequest'), (0, _param.default)('at', 'BlockHash', {
    isOptional: true
  })],
  type: 'ContractExecResult'
};
const getStorage = {
  description: 'Returns the value under a specified storage key in a contract',
  isOptional: true,
  params: [(0, _param.default)('address', 'AccountId'), (0, _param.default)('key', 'H256'), (0, _param.default)('at', 'BlockHash', {
    isOptional: true
  })],
  type: 'Option<Bytes>'
};
const section = 'contracts';
/**
 * @summary Calls to retrieve system info.
 */

var _default = {
  isDeprecated: false,
  isHidden: false,
  description: '(Optional) Methods that performs actions on contracts',
  section,
  methods: {
    call: (0, _method.default)(section, 'call', call),
    getStorage: (0, _method.default)(section, 'getStorage', getStorage)
  }
};
exports.default = _default;
