// @polkadot/jsonrpc/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _account = _interopRequireDefault(require("./account"));

var _author = _interopRequireDefault(require("./author"));

var _chain = _interopRequireDefault(require("./chain"));

var _contracts = _interopRequireDefault(require("./contracts"));

var _payment = _interopRequireDefault(require("./payment"));

var _rpc = _interopRequireDefault(require("./rpc"));

var _state = _interopRequireDefault(require("./state"));

var _system = _interopRequireDefault(require("./system"));

// Copyright 2017-2019 @polkadot/jsonrpc authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
const interfaces = {
  account: _account.default,
  author: _author.default,
  chain: _chain.default,
  contracts: _contracts.default,
  payment: _payment.default,
  rpc: _rpc.default,
  state: _state.default,
  system: _system.default
};
var _default = interfaces;
exports.default = _default;
