// @polkadot/jsonrpc/payment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _method = _interopRequireDefault(require("./create/method"));

var _param = _interopRequireDefault(require("./create/param"));

// Copyright 2017-2019 @polkadot/jsonrpc authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// NOTE order here is the same as in Rust, alphabetical below
const queryInfo = {
  description: 'Retrieves the fee information for an encoded extrinsic',
  isOptional: true,
  params: [(0, _param.default)('extrinsic', 'Bytes'), (0, _param.default)('at', 'BlockHash', {
    isOptional: true
  })],
  type: 'RuntimeDispatchInfo'
};
const section = 'payment';
/**
 * @summary Calls to retrieve payment information.
 */

var _default = {
  isDeprecated: false,
  isHidden: false,
  description: 'Methods that retrieves payment information, e.g. fee calculations',
  section,
  methods: {
    queryInfo: (0, _method.default)(section, 'queryInfo', queryInfo)
  }
};
exports.default = _default;
