// @polkadot/types/primitive/Extrinsic/v3/Extrinsic.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("@polkadot/util");

var _create = require("../../../codec/create");

var _Struct = _interopRequireDefault(require("../../../codec/Struct"));

var _ExtrinsicSignature = _interopRequireDefault(require("./ExtrinsicSignature"));

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
const TRANSACTION_VERSION = 3;

/**
 * @name ExtrinsicV3
 * @description
 * The third generation of compact extrinsics
 */
class ExtrinsicV3 extends _Struct.default {
  constructor(registry, value) {
    let {
      isSigned
    } = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    super(registry, {
      signature: _ExtrinsicSignature.default,
      method: 'Call'
    }, ExtrinsicV3.decodeExtrinsic(registry, value, isSigned));
  }

  static decodeExtrinsic(registry, value) {
    let isSigned = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

    if (!value) {
      return {};
    } else if (value instanceof ExtrinsicV3) {
      return value;
    } else if (value instanceof (0, _create.ClassOf)(registry, 'Call')) {
      return {
        method: value
      };
    } else if ((0, _util.isU8a)(value)) {
      // here we decode manually since we need to pull through the version information
      const signature = new _ExtrinsicSignature.default(registry, value, {
        isSigned
      });
      const method = (0, _create.createType)(registry, 'Call', value.subarray(signature.encodedLength));
      return {
        method,
        signature
      };
    }

    return value;
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    return this.toU8a().length;
  }
  /**
   * @description The [[Call]] this extrinsic wraps
   */


  get method() {
    return this.get('method');
  }
  /**
   * @description The [[ExtrinsicSignatureV3]]
   */


  get signature() {
    return this.get('signature');
  }
  /**
   * @description The version for the signature
   */


  get version() {
    return TRANSACTION_VERSION;
  }
  /**
   * @description Add an [[ExtrinsicSignatureV3]] to the extrinsic (already generated)
   */


  addSignature(signer, signature, payload) {
    this.signature.addSignature(signer, signature, payload);
    return this;
  }
  /**
   * @description Sign the extrinsic with a specific keypair
   */


  sign(account, options) {
    this.signature.sign(this.method, account, options);
    return this;
  }

}

exports.default = ExtrinsicV3;
