// datdot-explorer/src/node_modules/footer.js
const bel = require('bel')
const style_sheet = require('support-style-sheet')
const {i_button, i_link} = require('datdot-ui-button')
const i_dropdown = require('datdot-ui-dropdown')
const message_maker = require('message-maker')
const make_grid = require('make-grid')

module.exports = status_footer

function status_footer (opts, protocol) {
    const {page = '*', flow = 'i-footer', name, theme = {}} = opts
    const make = message_maker(`${name} / ${flow} / ${page}`)
    const recipients = []
    
    function widget () {
        const footer = document.createElement('i-footer')
        const shadow = footer.attachShadow({mode: 'open'})
        style_sheet(shadow, style)
        // const address_list = addresses.map( (url, index) => bel`
        // <li class="info">
        //   <span class="id">${index+1}</span>
        //   <span class="address">${url}</span>
        // </li>`)
        const theme_option = {
            message: {
                size: 'var(--size12)',
            },
            button: {
                // color: 'var(--color-black)',
                bg_color: 'var(--color-grey66)',
                padding: '2px 4px',
                border_radius: '0',
            }
        }
        const filter = bel`<input class="filter" type='text' name='filter' placeholder='Filter' aria-label='search filter'>`
        const expanded = i_button(
        {
            name: 'expanded', 
            body: 'Expanded', 
            role: 'switch',
            theme: {
                props: {
                    ...theme_option.button
                }
            }
        }, footer_protocol('expanded'))

        // option for terminal-selector 
        const terminal_option = 
        {
            page,
            name: 'terminal',
            mode : 'single-select',
            expanded: false,
            options: {
                button: {
                    name: 'terminal',
                    theme: {
                        props: {
                            border_radius: '0',
                            padding: '2px 4px',
                        }
                    }
                },
                list: {
                    name: 'terminal-selector',
                    array: [
                        {text: 'Compact messages'},
                        {text: 'Comfortable messages'}
                    ] 
                }
            },
        }

        const levels_option = 
        {
            page,
            name: 'levels',
            mode : 'single-select',
            expanded: false,
            options: {
                button: {
                    name: 'levels',
                    theme: {
                        props: {
                            border_radius: '0',
                            padding: '2px 4px',
                        }
                    }
                },
                list: {
                    name: 'levels-selector',
                    array: [
                        {text: 'Level1', current: true},
                        {text: 'Level2'},
                        {text: 'Level3'},
                        {text: 'Level4'},
                        {text: 'Level5'},
                    ] 
                }
            },
        }
        const terminal_selector = i_dropdown(terminal_option, footer_protocol(terminal_option.name))
        const levels_selector = i_dropdown(levels_option, footer_protocol(levels_option.name))
        const message_count = `1,972,710,231,626,326,235,437,253,235`
        const all_message = bel`<span class="total">All messages: ${message_count}</span>`
        const actions = bel`<div class="actions">${filter}${levels_selector}${terminal_selector}${expanded}${all_message}</div>`
        const status = bel`<div class="status">There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour</div>`
        shadow.append(actions, status)

        return footer

        // handle events
        function switch_event (from, data) {
            const state = !data
            recipients[from](make({type: 'switched', data: state}))
            send(make({to: from, type: 'triggered', data: {checked: state}}) )
            // todo: copy this to _old.js
            // send(make({to: 'terminal', type: 'layout-mode', data: {expanded: state}})) 
        }

        function selector_event (from, data) {
            const dropdowns = actions.querySelectorAll('i-dropdown')
            const state = data.expanded
            const type = state ? 'expanded' : 'collapsed'
            const to = `${from} / listbox / ui-list`
            recipients[from]( make({to, type, data: {from, expanded: state}}) )
            dropdowns.forEach( item => {
                const name = item.getAttribute('aria-label')
                const to = `${name} / listbox / ui-list`
                item.style.zIndex = '99'
                if (name !== from) {
                    recipients[name]( make({to, type: 'collapsed', data: {name, expanded: false}}) )
                    item.removeAttribute('style')
                }
            })
        }

        function click_event (from, role, data) {
            if (role === 'switch') return switch_event(from, data)
            if (role === 'listbox') return selector_event(from, data)
        }   
        function footer_protocol (name) {
            return send => {
                recipients[name] = send
                return get
            }
        }
        function get (msg) {
            const {head, type, data} = msg
            const from = head[0].split('/')[0].trim()
            const role = head[0].split(' / ')[1]
            if (type === 'click') return click_event (from, role, data)
        }
    }
    
    const style = `
        :host(i-footer) {
            --size: var(--size12);
            --color: var(--color-white);
            --bg-color: var(--color-dark);
            display: grid;
            font-size: var(--size);
            color: hsl(var(--color));
            background-color: hsl(var(--bg-color));
            ${make_grid({
                areas: ['actions', 'status']
            })}
            width: 100%;
            max-width: 100%;
        }

        .actions {
            grid-area: actions;
            display: grid;
            ${make_grid({
                rows: 'minmax(0, 30px) auto',
                columns: 'minmax(120px, 300px) 90px 155px 100px auto',
                gap: '6px'
            })}
            padding: 6px;
        }
        .total {
            
        }
        .filter {
        }
        .status {
            grid-area: status;
            padding: 0 8px 8px;
        }
    `
    return widget()
}
