// datdot-explorer/src/node_modules/footer.js
const bel = require('bel')
const style_sheet = require('support-style-sheet')
const message_maker = require('message-maker')
const make_grid = require('make-grid')

module.exports = footer

function footer (opts, protocol) {
    const {page = '*', flow = 'i-status', name, theme = {}} = opts
    const make = message_maker(`${name} / ${flow} / ${page}`)
    const recipients = []

    function widget () {
        const send = protocol(get)
        const el = document.createElement('i-footer')
        const shadow = el.attachShadow({mode: 'open'})
        send(make({type: 'ready'}))
        style_sheet(shadow, style)
        // const address_list = addresses.map( (url, index) => bel`
        // <li class="info">
        //   <span class="id">${index+1}</span>
        //   <span class="address">${url}</span>
        // </li>`)

        const status = bel`<div class="status">There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour</div>`
        shadow.append(status)

        return el

        function get (msg) {
            const {head, type, data} = msg
            const from = head[0].split('/')[0].trim()
            const role = head[0].split(' / ')[1]
        }
    }
    
    const style = `
        :host(i-footer) {
            --size: var(--size12);
            --color: var(--color-white);
            --bg-color: var(--color-dark);
            display: grid;
            font-size: var(--size);
            color: hsl(var(--color));
            background-color: hsl(var(--bg-color));
            ${make_grid({
                areas: ['actions', 'status']
            })}
            width: 100%;
            max-width: 100%;
        }

        .actions {
            grid-area: actions;
            display: grid;
            ${make_grid({
                rows: 'minmax(0, 30px) auto',
                columns: 'minmax(120px, 300px) 90px minmax(155px, 175px) 100px auto',
                gap: '6px'
            })}
            padding: 6px;
        }
        .total {
            
        }
        .filter {
        }
        .status {
            grid-area: status;
            padding: 0 8px 8px;
        }
    `
    return widget()
}
