// @polkadot/types/primitive/IdentityFields.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bn = _interopRequireDefault(require("bn.js"));

var _util = require("@polkadot/util");

var _Set = _interopRequireDefault(require("../codec/Set"));

var _createType = require("../codec/create/createType");

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
function decodeFields(registry, value) {
  if ((0, _util.isString)(value)) {
    return decodeFields(registry, (0, _util.u8aToU8a)(value.toString()));
  } else if ((0, _util.isU8a)(value)) {
    return (0, _createType.createType)(registry, 'u64', value);
  }

  return value;
}

class IdentityFields extends _Set.default {
  constructor(registry, value) {
    super(registry, {
      Display: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000000001),
      Legal: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000000010),
      Web: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000000100),
      Riot: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000001000),
      Email: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000010000),
      PgpFingerprint: new _bn.default(0b0000000000000000000000000000000000000000000000000000000000100000),
      Image: new _bn.default(0b0000000000000000000000000000000000000000000000000000000001000000)
    }, decodeFields(registry, value));
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    return 8;
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   */


  toU8a() {
    return (0, _createType.createType)(this.registry, 'u64', this.valueEncoded).toU8a();
  }

}

exports.default = IdentityFields;
