// @polkadot/types/codec/Set.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bn = _interopRequireDefault(require("bn.js"));

var _util = require("@polkadot/util");

var _utilCrypto = require("@polkadot/util-crypto");

var _Raw = _interopRequireDefault(require("./Raw"));

var _utils = require("./utils");

// Copyright 2017-2020 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.

/**
 * @name Set
 * @description
 * An Set is an array of string values, represented an an encoded type by
 * a bitwise representation of the values.
 */
// FIXME This is a prime candidate to extend the JavaScript built-in Set
class CodecSet extends Set {
  constructor(registry, setValues, value) {
    super(CodecSet.decodeSet(setValues, value));
    this.registry = void 0;
    this._setValues = void 0;
    this.registry = registry;
    this._setValues = setValues;
  }

  static decodeSet(setValues, value = 0) {
    if ((0, _util.isString)(value)) {
      return CodecSet.decodeSet(setValues, (0, _util.u8aToU8a)(value));
    } else if ((0, _util.isU8a)(value)) {
      return value.length === 0 ? [] : CodecSet.decodeSetNumber(setValues, value[0]);
    } else if (value instanceof Set) {
      return CodecSet.decodeSetArray(setValues, [...value.values()]);
    } else if (Array.isArray(value)) {
      return CodecSet.decodeSetArray(setValues, value);
    }

    return CodecSet.decodeSetNumber(setValues, value);
  }

  static decodeSetArray(setValues, value) {
    return value.reduce((result, key) => {
      (0, _util.assert)(!(0, _util.isUndefined)(setValues[key]), "Set: Invalid key '".concat(key, "' passed to Set, allowed ").concat(Object.keys(setValues).join(', ')));
      result.push(key);
      return result;
    }, []);
  }

  static decodeSetNumber(setValues, _value) {
    const bn = (0, _util.bnToBn)(_value);
    const result = Object.keys(setValues).reduce((result, key) => {
      if (bn.and((0, _util.bnToBn)(setValues[key])).eq((0, _util.bnToBn)(setValues[key]))) {
        result.push(key);
      }

      return result;
    }, []);
    const computed = CodecSet.encodeSet(setValues, result);
    (0, _util.assert)(bn.eq(computed), "Set: Mismatch decoding '".concat(bn, "', computed as '").concat(computed, "' with ").concat(result));
    return result;
  }

  static encodeSet(setValues, value) {
    return value.reduce((result, value) => {
      return result.or((0, _util.bnToBn)(setValues[value] || 0));
    }, new _bn.default(0));
  }

  static with(values) {
    return class extends CodecSet {
      constructor(registry, value) {
        super(registry, values, value);
        Object.keys(values).forEach(_key => {
          const name = (0, _util.stringUpperFirst)((0, _util.stringCamelCase)(_key));
          const iskey = "is".concat(name); // do not clobber existing properties on the object

          if ((0, _util.isUndefined)(this[iskey])) {
            Object.defineProperty(this, iskey, {
              enumerable: true,
              get: () => this.strings.includes(_key)
            });
          }
        });
      }

    };
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    return 1;
  }
  /**
   * @description returns a hash of the contents
   */


  get hash() {
    return new _Raw.default(this.registry, (0, _utilCrypto.blake2AsU8a)(this.toU8a(), 256));
  }
  /**
   * @description true is the Set contains no values
   */


  get isEmpty() {
    return this.size === 0;
  }
  /**
   * @description The actual set values as a string[]
   */


  get strings() {
    return [...super.values()];
  }
  /**
   * @description The encoded value for the set members
   */


  get valueEncoded() {
    return CodecSet.encodeSet(this._setValues, this.strings);
  }
  /**
   * @description adds a value to the Set (extended to allow for validity checking)
   */


  add(key) {
    // we have the isUndefined(this._setValues) in here as well, add is used internally
    // in the Set constructor (so it is undefined at this point, and should allow)
    (0, _util.assert)((0, _util.isUndefined)(this._setValues) || !(0, _util.isUndefined)(this._setValues[key]), "Set: Invalid key '".concat(key, "' on add"));
    super.add(key);
    return this;
  }
  /**
   * @description Compares the value of the input to see if there is a match
   */


  eq(other) {
    if (Array.isArray(other)) {
      // we don't actually care about the order, sort the values
      return (0, _utils.compareArray)(this.strings.sort(), other.sort());
    } else if (other instanceof Set) {
      return this.eq([...other.values()]);
    } else if ((0, _util.isNumber)(other) || (0, _util.isBn)(other)) {
      return this.valueEncoded.eq((0, _util.bnToBn)(other));
    }

    return false;
  }
  /**
   * @description Returns a hex string representation of the value
   */


  toHex() {
    return (0, _util.u8aToHex)(this.toU8a());
  }
  /**
   * @description Converts the Object to JSON, typically used for RPC transfers
   */


  toJSON() {
    return this.strings;
  }
  /**
   * @description The encoded value for the set members
   */


  toNumber() {
    return this.valueEncoded.toNumber();
  }
  /**
   * @description Returns the base runtime type name for this instance
   */


  toRawType() {
    // FIXME We don't cater for this in createType as of yet
    return JSON.stringify({
      _set: this._setValues
    });
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return "[".concat(this.strings.join(', '), "]");
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   * @param isBare true when the value has none of the type-specific prefixes (internal)
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars


  toU8a(isBare) {
    return new Uint8Array([this.valueEncoded.toNumber()]);
  }

}

exports.default = CodecSet;
