// @polkadot/metadata/Metadata/v3/toV4.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = toV4;

var _codec = require("@polkadot/types/codec");

// Copyright 2017-2020 @polkadot/metadata authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.

/**
 * Convert V3 StorageFunction to V4 StorageFunction
 */
function toV4StorageFunction(registry, storageFn) {
  const {
    documentation,
    fallback,
    modifier,
    name,
    type
  } = storageFn; // Convert the old type to the new type: there is one new field
  // called `hasher`, which we initialize to xxHash (the default in
  // v3).

  const [newType, index] = type.isPlain ? [type.asPlain, 0] : type.isMap ? [(0, _codec.createType)(registry, 'MapTypeV4', {
    hasher: (0, _codec.createType)(registry, 'StorageHasherV4', 'Twox128'),
    key: type.asMap.key,
    value: type.asMap.value,
    linked: type.asMap.linked
  }), 1] : [(0, _codec.createType)(registry, 'DoubleMapTypeV4', {
    hasher: (0, _codec.createType)(registry, 'StorageHasherV4', 'Twox128'),
    key1: type.asDoubleMap.key1,
    key2: type.asDoubleMap.key2,
    value: type.asDoubleMap.value,
    key2Hasher: type.asDoubleMap.key2Hasher
  }), 2];
  return (0, _codec.createType)(registry, 'StorageFunctionMetadataV4', {
    documentation,
    fallback,
    name,
    modifier,
    type: (0, _codec.createType)(registry, 'StorageFunctionTypeV4', newType, index)
  });
}
/**
 * Convert from MetadataV3 to MetadataV4
 * See https://github.com/paritytech/substrate/pull/2268 for details
 */


function toV4(registry, {
  modules
}) {
  return (0, _codec.createType)(registry, 'MetadataV4', {
    modules: modules.map(({
      calls,
      events,
      name,
      prefix,
      storage
    }) => (0, _codec.createType)(registry, 'ModuleMetadataV4', {
      calls,
      events,
      name,
      prefix,
      storage: storage.isSome ? new _codec.Option(registry, 'Vec<StorageFunctionMetadataV4>', storage.unwrap().map(v => toV4StorageFunction(registry, v))) : undefined
    }))
  });
}
