// @polkadot/api/base/Decorate.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bn = _interopRequireDefault(require("bn.js"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _apiDerive = _interopRequireDefault(require("@polkadot/api-derive"));

var _Decorated = _interopRequireDefault(require("@polkadot/metadata/Decorated"));

var _getHasher = _interopRequireDefault(require("@polkadot/metadata/Decorated/storage/fromMetadata/getHasher"));

var _rpcCore = _interopRequireDefault(require("@polkadot/rpc-core"));

var _rpcProvider = require("@polkadot/rpc-provider");

var _types = require("@polkadot/types");

var _Linkage = require("@polkadot/types/codec/Linkage");

var _constants = require("@polkadot/types/extrinsic/constants");

var _util = require("@polkadot/util");

var _submittable = require("../submittable");

var _decorate = require("../util/decorate");

var _augmentObject = _interopRequireDefault(require("../util/augmentObject"));

var _Events = _interopRequireDefault(require("./Events"));

// Copyright 2017-2020 @polkadot/api authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class Decorate extends _Events.default {
  // HACK Use BN import so decorateDerive works... yes, wtf.

  /**
   * This is the one and only method concrete children classes need to implement.
   * It's a higher-order function, which takes one argument
   * `method: Method extends (...args: any[]) => Observable<any>`
   * (and one optional `options`), and should return the user facing method.
   * For example:
   * - For ApiRx, `decorateMethod` should just be identity, because the input
   * function is already an Observable
   * - For ApiPromise, `decorateMethod` should return a function that takes all
   * the parameters from `method`, adds an optional `callback` argument, and
   * returns a Promise.
   *
   * We could easily imagine other user-facing interfaces, which are simply
   * implemented by transforming the Observable to Stream/Iterator/Kefir/Bacon
   * via `decorateMethod`.
   */

  /**
   * @description Create an instance of the class
   *
   * @param options Options object to create API instance or a Provider instance
   *
   * @example
   * <BR>
   *
   * ```javascript
   * import Api from '@polkadot/api/promise';
   *
   * const api = new Api().isReady();
   *
   * api.rpc.subscribeNewHeads((header) => {
   *   console.log(`new block #${header.number.toNumber()}`);
   * });
   * ```
   */
  constructor(options, type, decorateMethod) {
    super();
    this.registry = void 0;
    this.__phantom = new _bn.default(0);
    this._consts = {};
    this._derive = void 0;
    this._extrinsics = void 0;
    this._extrinsicType = _constants.DEFAULT_VERSION;
    this._genesisHash = void 0;
    this._isConnected = void 0;
    this._isReady = false;
    this._options = void 0;
    this._query = {};
    this._queryMulti = void 0;
    this._rpc = void 0;
    this._rpcCore = void 0;
    this._runtimeChain = void 0;
    this._runtimeMetadata = void 0;
    this._runtimeVersion = void 0;
    this._rx = {
      consts: {},
      query: {},
      tx: {}
    };
    this._type = void 0;
    this.decorateMethod = void 0;

    this.rxDecorateMethod = method => {
      return method;
    };

    this.registry = options.registry || new _types.TypeRegistry();
    const thisProvider = options.source ? options.source._rpcCore.provider.clone() : options.provider || new _rpcProvider.WsProvider();
    this.decorateMethod = decorateMethod;
    this._options = options;
    this._type = type;
    this._rpcCore = new _rpcCore.default(this.registry, thisProvider, this._options.rpc);
    this._isConnected = new _rxjs.BehaviorSubject(this._rpcCore.provider.isConnected());
    this._rx.hasSubscriptions = this._rpcCore.provider.hasSubscriptions;
    this._rx.registry = this.registry;
  }
  /**
   * @description Creates an instance of a type as registered
   */


  createType(type, ...params) {
    return this.registry.createType(type, ...params);
  }
  /**
   * @description Register additional user-defined of chain-specific types in the type registry
   */


  registerTypes(types) {
    types && this.registry.register(types);
  }
  /**
   * @returns `true` if the API operates with subscriptions
   */


  get hasSubscriptions() {
    return this._rpcCore.provider.hasSubscriptions;
  }

  injectMetadata(metadata, fromEmpty) {
    const decoratedMeta = new _Decorated.default(this.registry, metadata);

    if (fromEmpty || !this._extrinsics) {
      this._extrinsics = this.decorateExtrinsics(decoratedMeta.tx, this.decorateMethod);
      this._rx.tx = this.decorateExtrinsics(decoratedMeta.tx, this.rxDecorateMethod);
    } else {
      (0, _augmentObject.default)('tx', this.decorateExtrinsics(decoratedMeta.tx, this.decorateMethod), this._extrinsics, false);
      (0, _augmentObject.default)('', this.decorateExtrinsics(decoratedMeta.tx, this.rxDecorateMethod), this._rx.tx, false);
    } // this API


    (0, _augmentObject.default)('query', this.decorateStorage(decoratedMeta.query, this.decorateMethod), this._query, fromEmpty);
    (0, _augmentObject.default)('consts', decoratedMeta.consts, this._consts, fromEmpty); // rx

    (0, _augmentObject.default)('', this.decorateStorage(decoratedMeta.query, this.rxDecorateMethod), this._rx.query, fromEmpty);
    (0, _augmentObject.default)('', decoratedMeta.consts, this._rx.consts, fromEmpty);
  }

  decorateFunctionMeta(input, output) {
    output.meta = input.meta;
    output.method = input.method;
    output.section = input.section;
    output.toJSON = input.toJSON;

    if (input.callIndex) {
      output.callIndex = input.callIndex;
    }

    return output;
  } // Filter all RPC methods based on the results of the rpc_methods call. We do this in the following
  // manner to cater for both old and new:
  //   - when the number of entries are 0, only remove the ones with isOptional (account & contracts)
  //   - when non-zero, remove anything that is not in the array (we don't do this)


  async filterRpc() {
    let methods;

    try {
      // we ignore the version (adjust as versions change, for now only "1")
      methods = (await this._rpcCore.rpc.methods().toPromise()).methods.map(t => t.toString());
    } catch (error) {
      // the method is not there, we adjust accordingly
      methods = [];
    }

    this.filterRpcMethods(methods);
  }

  filterRpcMethods(methods) {
    // this is true when the RPC has entries
    const hasResults = methods.length !== 0; // loop through all entries we have (populated in decorate) and filter as required

    [...this._rpcCore.mapping.entries()].filter(([key, {
      isOptional
    }]) => // only remove when we have results and method missing, or with no results if optional
    hasResults ? !methods.includes(key) && key !== 'rpc_methods' // rpc_methods doesn't appear, v1
    : isOptional || key === 'rpc_methods' // we didn't find this one, remove
    ) // eslint-disable-next-line @typescript-eslint/no-unused-vars
    .forEach(([_, {
      method,
      section
    }]) => {
      delete this._rpc[section][method];
      delete this._rx.rpc[section][method];
    });
  }

  decorateRpc(rpc, decorateMethod) {
    return rpc.sections.reduce((out, _sectionName) => {
      const sectionName = _sectionName; // out and section here are horrors to get right from a typing perspective :(

      out[sectionName] = Object.entries(rpc[sectionName]).reduce((section, [methodName, method]) => {
        //  skip subscriptions where we have a non-subscribe interface
        if (this.hasSubscriptions || !(methodName.startsWith('subscribe') || methodName.startsWith('unsubscribe'))) {
          section[methodName] = decorateMethod(method, {
            methodName
          });
        }

        return section;
      }, {});
      return out;
    }, {});
  }

  decorateMulti(decorateMethod) {
    return decorateMethod(calls => this._rpcCore.state.subscribeStorage(calls.map(arg => // the input is a QueryableStorageEntry, convert to StorageEntry
    Array.isArray(arg) ? [arg[0].creator, ...arg.slice(1)] : [arg.creator])));
  }

  decorateExtrinsics(extrinsics, decorateMethod) {
    const creator = (0, _submittable.createSubmittable)(this._type, this._rx, decorateMethod);
    return Object.entries(extrinsics).reduce((out, [name, section]) => {
      out[name] = Object.entries(section).reduce((out, [name, method]) => {
        out[name] = this.decorateExtrinsicEntry(method, creator);
        return out;
      }, {});
      return out;
    }, creator);
  }

  decorateExtrinsicEntry(method, creator) {
    const decorated = (...params) => creator(method(...params));

    return this.decorateFunctionMeta(method, decorated);
  }

  decorateStorage(storage, decorateMethod) {
    return Object.entries(storage).reduce((out, [name, section]) => {
      out[name] = Object.entries(section).reduce((out, [name, method]) => {
        out[name] = this.decorateStorageEntry(method, decorateMethod);
        return out;
      }, {});
      return out;
    }, {});
  }

  decorateStorageEntry(creator, decorateMethod) {
    // get the storage arguments, with DoubleMap as an array entry, otherwise spread
    const getArgs = (...args) => creator.meta.type.isDoubleMap ? [creator, args] : [creator, ...args]; // FIXME We probably want to be able to query the full list with non-subs as well


    const decorated = this.hasSubscriptions && creator.iterKey && creator.meta.type.isMap && creator.meta.type.asMap.linked.isTrue ? this.decorateStorageLinked(creator, decorateMethod) : decorateMethod((...args) => this.hasSubscriptions ? this._rpcCore.state // Unfortunately for one-shot calls we also use .subscribeStorage here
    .subscribeStorage([getArgs(...args)]) // state_storage returns an array of values, since we have just subscribed to
    // a single entry, we pull that from the array and return it as-is
    .pipe((0, _operators.map)(([data]) => data)) : this._rpcCore.state.getStorage(getArgs(...args)), {
      methodName: creator.method
    });
    decorated.creator = creator;
    decorated.at = decorateMethod((hash, arg1, arg2) => this._rpcCore.state.getStorage(getArgs(arg1, arg2), hash));
    decorated.entries = decorateMethod(doubleMapArg => this.retrieveMapEntries(creator, doubleMapArg));
    decorated.hash = decorateMethod((arg1, arg2) => this._rpcCore.state.getStorageHash(getArgs(arg1, arg2)));

    decorated.key = (arg1, arg2) => (0, _util.u8aToHex)((0, _util.compactStripLength)(creator(creator.meta.type.isDoubleMap ? [arg1, arg2] : arg1))[1]);

    decorated.keyPrefix = () => (0, _util.u8aToHex)(creator.keyPrefix);

    if (this.hasSubscriptions) {
      // When using double map storage function, user need to pass double map key as an array
      decorated.multi = decorateMethod(args => this._rpcCore.state.subscribeStorage(args.map(arg => [creator, arg])));
    }

    decorated.size = decorateMethod((arg1, arg2) => this._rpcCore.state.getStorageSize(getArgs(arg1, arg2)));
    return this.decorateFunctionMeta(creator, decorated);
  }

  decorateStorageLinked(creator, decorateMethod) {
    const result = new Map();
    let subject;
    let head = null; // retrieve a value based on the key, iterating if it has a next entry. Since
    // entries can be re-linked in the middle of a list, we subscribe here to make
    // sure we catch any updates, no matter the list position

    const getNext = key => this._rpcCore.state.subscribeStorage([[creator, key]]).pipe((0, _operators.switchMap)(([_data]) => {
      const data = creator.meta.modifier.isOptional ? _data.unwrapOr(null) : _data;
      result.set(key, data); // iterate from this key to the linkages, constructing entries for all
      // those found and available

      if (data && data[1].next.isSome) {
        return getNext(data[1].next.unwrap());
      }

      const [keys, vals] = [[], []];
      let nextKey = head; // loop through the results collected, starting at the head an re-creating
      // the list. Our map may have old entries, based on the linking these will
      // not be returned in the final result

      while (nextKey) {
        var _linkage$next;

        const entry = result.get(nextKey);

        if (!entry) {
          break;
        }

        const [item, linkage] = entry;
        keys.push(nextKey);
        vals.push(item);
        nextKey = (_linkage$next = linkage.next) === null || _linkage$next === void 0 ? void 0 : _linkage$next.unwrapOr(null);
      }

      const nextResult = vals.length ? new _Linkage.LinkageResult(this.registry, [keys[0].constructor, keys], [vals[0].constructor, vals]) : new _Linkage.LinkageResult(this.registry, [_types.Null, []], [_types.Null, []]); // we set our result into a subject so we have a single observable for
      // which the value changes over time. Initially create, follow-up next

      if (subject) {
        subject.next(nextResult);
      } else {
        subject = new _rxjs.BehaviorSubject(nextResult);
      }

      return subject;
    })); // this handles the case where the head changes effectively, i.e. a new entry
    // appears at the top of the list, the new getNext gets kicked off


    return decorateMethod((...args) => args.length ? this._rpcCore.state.subscribeStorage([[creator, ...args]]).pipe((0, _operators.map)(([data]) => data)) : this._rpcCore.state.subscribeStorage([creator.iterKey]).pipe((0, _operators.switchMap)(([key]) => getNext(head = key))));
  }

  retrieveMapEntries(creator, arg) {
    (0, _util.assert)(creator.iterKey && (creator.meta.type.isMap || creator.meta.type.isDoubleMap), 'entries can only be retrieved on maps, linked maps and double maps');
    const outputType = creator.meta.type.isMap ? creator.meta.type.asMap.value.toString() : creator.meta.type.asDoubleMap.value.toString();
    return this._rpcCore.state // FIXME This should really be some sort of subscription, so we can get stuff as
    // it changes (as of now it is a one-shot query). Not sure how to do this though...
    .getKeys(this.createType('Raw', (0, _util.u8aConcat)(creator.iterKey, creator.meta.type.isDoubleMap && !(0, _util.isUndefined)(arg) && !(0, _util.isNull)(arg) ? (0, _getHasher.default)(creator.meta.type.asDoubleMap.hasher)(this.createType(creator.meta.type.asDoubleMap.key1.toString(), arg).toU8a()) : new Uint8Array([])))).pipe((0, _operators.switchMap)(keys => (0, _rxjs.combineLatest)([(0, _rxjs.of)(keys), // Not 100% sure about this, surely this is just a vec?
    this._rpcCore.state.subscribeStorage(keys.map(key => key.setOutputType(outputType)))])), (0, _operators.map)(([keys, values]) => keys.map((key, index) => [key, values[index]])));
  }

  decorateDeriveRx(decorateMethod) {
    // Pull in derive from api-derive
    const derive = (0, _apiDerive.default)(this._rx, this._options.derives);
    return (0, _decorate.decorateSections)(derive, decorateMethod);
  }

  decorateDerive(decorateMethod) {
    return (0, _decorate.decorateSections)(this._rx.derive, decorateMethod);
  }
  /**
   * Put the `this.onCall` function of ApiRx here, because it is needed by
   * `api._rx`.
   */


}

exports.default = Decorate;
