// datdot-ui-button/demo/node_modules/img-btn.js
module.exports = img_btn

function img_btn ({name, body, icon = {}, cover, disabled, props = {}}, button, protocol) {
    var {icon_name = 'edit', path, classlist} = icon
    const {
        icon_size = '20px',
        icon_fill = 'var(--primary-icon-fill)',
        icon_fill_hover = 'var(--primary-icon-fill-hover)',
        avatar_width = '150px', 
        avatar_height = '150px',
        weight = '600', 
        size = 'var(--size24)', 
        size_hover = 'var(--size36)', 
        color = 'var(--color-amaranth-pink)', 
        color_hover = 'var(--primary-color-hover)',
        bg_color = 'var(--color-white)',
        bg_color_hover = 'var(--color-grey88)', 
        border_radius = '8px',
        avatar_radius = '50%',
        disabled_size = 'var(--size24)',
        shadow_opacity = '0.2'
    } = props
    return button(
        {
            name, 
            body,
            icons: {
                icon: {name: icon_name, path}
            },
            cover,
            classlist,
            disabled,
            theme:
            { 
                // to solve border-radius & overflow: hidden not working on safari
                // -webkit-mask-image: -webkit-radial-gradient(white, black)
                // transform: translateZ(0)
                style: `
                :host(i-button) img {
                    object-position: 0 -20px
                }
                `, 
                props: { icon_size, icon_fill, icon_fill_hover, avatar_width, avatar_height, weight, size, size_hover, color, color_hover, bg_color, bg_color_hover, border_radius, avatar_radius, disabled_size, shadow_opacity},
                grid: {
                    button: {
                        // areas: ['icon text', 'avatar avatar'],
                        rows: 'auto 1fr',
                        columns: 'repeat(auto-fill, minmax(0, auto)) auto',
                        justify: 'items-center'
                    },
                    icon: {
                        // area: 'icon',
                        row: '1',
                        column: '1',
                        justify: 'self-right'
                    },
                    avatar: {
                        // area: 'avatar',
                        row: '2',
                        column: 'span 3',
                        // justify: 'self-center'
                    },
                    text: {
                        // area: 'text',
                        row: '1',
                        column: '2 / span 2',
                        justify: 'self-left'

                    }
                }
            }
        }, protocol)
}
