// @polkadot/types/primitive/Generic/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GenericAccountId", {
  enumerable: true,
  get: function get() {
    return _AccountId.default;
  }
});
Object.defineProperty(exports, "GenericAccountIndex", {
  enumerable: true,
  get: function get() {
    return _AccountIndex.default;
  }
});
Object.defineProperty(exports, "GenericAddress", {
  enumerable: true,
  get: function get() {
    return _Address.default;
  }
});
Object.defineProperty(exports, "GenericBlock", {
  enumerable: true,
  get: function get() {
    return _Block.default;
  }
});
Object.defineProperty(exports, "GenericCall", {
  enumerable: true,
  get: function get() {
    return _Call.default;
  }
});
Object.defineProperty(exports, "GenericConsensusEngineId", {
  enumerable: true,
  get: function get() {
    return _ConsensusEngineId.default;
  }
});
Object.defineProperty(exports, "GenericEvent", {
  enumerable: true,
  get: function get() {
    return _Event.default;
  }
});
Object.defineProperty(exports, "GenericDigest", {
  enumerable: true,
  get: function get() {
    return _Digest.default;
  }
});
Object.defineProperty(exports, "GenericDigestItem", {
  enumerable: true,
  get: function get() {
    return _Digest.DigestItem;
  }
});
Object.defineProperty(exports, "GenericOrigin", {
  enumerable: true,
  get: function get() {
    return _Origin.default;
  }
});
Object.defineProperty(exports, "GenericVote", {
  enumerable: true,
  get: function get() {
    return _Vote.default;
  }
});

var _AccountId = _interopRequireDefault(require("./AccountId"));

var _AccountIndex = _interopRequireDefault(require("./AccountIndex"));

var _Address = _interopRequireDefault(require("./Address"));

var _Block = _interopRequireDefault(require("./Block"));

var _Call = _interopRequireDefault(require("./Call"));

var _ConsensusEngineId = _interopRequireDefault(require("./ConsensusEngineId"));

var _Event = _interopRequireDefault(require("./Event"));

var _Digest = _interopRequireWildcard(require("./Digest"));

var _Origin = _interopRequireDefault(require("./Origin"));

var _Vote = _interopRequireDefault(require("./Vote"));
